node_monitor/node_provider_db.py

```diff
@@ -150,6 +151,14 @@ def _execute1(self, sql: str, params: Tuple[Any, ...]) -> List[Tuple[Any, ...]]:
         return result
     
 
+    def execute_write(self, sql: str, params: Tuple[Any, ...]) -> None:
+        conn = self.pool.getconn()
+        with conn.cursor() as cur:
+            cur.execute(sql, params)
+        conn.commit()
+        self.pool.putconn(conn)
+
+
```

Comment:
I haven't tried to write any data using `_execute` or `_execute1` but I assume they both should work fine. Is there a reason why you decided to reimplement these?

Reply:
this is because `_execute` and `_execute1` expect a result. So if you try to write using these functions, it will fail as soon as we get to the line `cur.fetchall`, as insert, update or delete functions don't typically return a result.

I propose we rename `_execute` and `_execute1` to `_execute_read` and `_execute_read1`. Thoughts?

Reply:
I'd rather just use one function, I think it keeps things simpler. There's got to be a way to get it to work agnostically.

ChatGPT gave me these. Maybe one of these works for our use case:

```py
def _execute(self, sql: str, params: Tuple[Any, ...]) -> List[Dict[str, Any]]:
    """Execute a SQL statement with a connection from the pool.
    An empty tuple should be passed if no parameters are needed.
    All transactions are committed.
    Returns a list of dicts instead of the default list of tuples.
    Ex. [{'column_name': value, ...}, ...]
    """
    conn = self.pool.getconn()
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute(sql, params)
        
        # Use try/except to handle fetchall() for non-SELECT operations
        try:
            result = cur.fetchall()
        except psycopg2.ProgrammingError:
            result = []
            
    conn.commit()
    self.pool.putconn(conn)
    
    return [dict(r) for r in result]

```
OR

```py
def _execute(self, sql: str, params: Tuple[Any, ...]) -> List[Dict[str, Any]]:
    """Execute a SQL statement with a connection from the pool.
    An empty tuple should be passed if no parameters are needed.
    All transactions are committed.
    Returns a list of dicts instead of the default list of tuples.
    Ex. [{'column_name': value, ...}, ...]
    """
    conn = self.pool.getconn()
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute(sql, params)
        
        # Check if there are results available
        if cur.rowcount > 0:
            result = [dict(r) for r in cur.fetchall()]
        else:
            result = []
            
    conn.commit()
    self.pool.putconn(conn)
    
    return result

```

Reply:
2nd method doesn't catch the case we are trying to handle. I still get the "no results to fetch error"

Instead I am using 
```py
conn = self.pool.getconn()
        with conn.cursor(cursor_factory=RealDictCursor) as cur:
            cur.execute(sql, params)
            
            # Check query is a SELECT statement by seeing if it returned results
            if cur.description is not None:
                result = [dict(r) for r in cur.fetchall()]
            else:
                result = []
                
        conn.commit()
        self.pool.putconn(conn)
        
        return result
```
If the description is none, then that means there were no results returned, which is typical for INSERT, UPDATE or DELETE statements.